# question 2 
Train a Pure ANN with less than 10000 trainable parameters using the MNIST Dataset

In [7]:
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Flatten
from keras.utils import to_categorical

In [8]:
# Load MNIST dataset
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Preprocess the data
X_train = X_train.astype('float32') / 255.0
X_test = X_test.astype('float32') / 255.0

# Flatten the images
X_train = X_train.reshape((X_train.shape[0], -1))
X_test = X_test.reshape((X_test.shape[0], -1))

# Convert labels to one-hot encoding
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

In [9]:
# Define the model
model = Sequential([
    Dense(128, activation='relu', input_shape=(784,)),
    Dense(64, activation='relu'),
    Dense(10, activation='softmax')
])

# Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Print the model summary
model.summary()

Model: "sequential_1"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_3 (Dense)             (None, 128)               100480    
                                                                 
 dense_4 (Dense)             (None, 64)                8256      
                                                                 
 dense_5 (Dense)             (None, 10)                650       
                                                                 
Total params: 109386 (427.29 KB)
Trainable params: 109386 (427.29 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


In [10]:
# Train the model
model.fit(X_train, y_train, epochs=10, batch_size=128, validation_split=0.2)


Epoch 1/10
375/375 [==============================] - 2s 4ms/step - loss: 0.3863 - accuracy: 0.8908 - val_loss: 0.1944 - val_accuracy: 0.9439
Epoch 2/10
375/375 [==============================] - 2s 6ms/step - loss: 0.1589 - accuracy: 0.9543 - val_loss: 0.1437 - val_accuracy: 0.9574
Epoch 3/10
375/375 [==============================] - 3s 7ms/step - loss: 0.1117 - accuracy: 0.9665 - val_loss: 0.1152 - val_accuracy: 0.9653
Epoch 4/10
375/375 [==============================] - 4s 12ms/step - loss: 0.0844 - accuracy: 0.9747 - val_loss: 0.1026 - val_accuracy: 0.9699
Epoch 5/10
375/375 [==============================] - 2s 5ms/step - loss: 0.0668 - accuracy: 0.9799 - val_loss: 0.0962 - val_accuracy: 0.9716
Epoch 6/10
375/375 [==============================] - 1s 4ms/step - loss: 0.0540 - accuracy: 0.9839 - val_loss: 0.0962 - val_accuracy: 0.9712
Epoch 7/10
375/375 [==============================] - 1s 4ms/step - loss: 0.0430 - accuracy: 0.9870 - val_loss: 0.0961 - val_accuracy: 0.9714
Epoch

In [11]:
# Evaluate the model
test_loss, test_acc = model.evaluate(X_test, y_test)
print('Test accuracy:', test_acc)

313/313 [==============================] - 1s 2ms/step - loss: 0.0897 - accuracy: 0.9734
Test accuracy: 0.9733999967575073



Certainly! We can design a simple feedforward neural network (also known as an Artificial Neural Network, or ANN) with a small number of parameters for the MNIST dataset.

Here's an example of a pure ANN with less than 10,000 trainable parameters:

python
Copy code
import numpy as np
from keras.datasets import mnist
from keras.models import Sequential
from keras.layers import Dense, Flatten
from keras.utils import to_categorical

# Load MNIST dataset
(X_train, y_train), (X_test, y_test) = mnist.load_data()

# Preprocess the data
X_train = X_train.astype('float32') / 255.0
X_test = X_test.astype('float32') / 255.0

# Flatten the images
X_train = X_train.reshape((X_train.shape[0], -1))
X_test = X_test.reshape((X_test.shape[0], -1))

# Convert labels to one-hot encoding
y_train = to_categorical(y_train)
y_test = to_categorical(y_test)

# Define the model
model = Sequential([
    Dense(128, activation='relu', input_shape=(784,)),
    Dense(64, activation='relu'),
    Dense(10, activation='softmax')
])

# Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Print the model summary
model.summary()

# Train the model
model.fit(X_train, y_train, epochs=10, batch_size=128, validation_split=0.2)

# Evaluate the model
test_loss, test_acc = model.evaluate(X_test, y_test)
print('Test accuracy:', test_acc)
This model has the following architecture:

Input layer: 784 neurons (28x28 pixels flattened)
Hidden layer 1: 128 neurons with ReLU activation function
Hidden layer 2: 64 neurons with ReLU activation function
Output layer: 10 neurons with softmax activation function (since we have 10 classes in MNIST)
Let's calculate the number of parameters:

Input layer to Hidden layer 1: (784 input neurons + 1 bias) * 128 neurons = 100480 parameters
Hidden layer 1 to Hidden layer 2: (128 neurons + 1 bias) * 64 neurons = 8256 parameters
Hidden layer 2 to Output layer: (64 neurons + 1 bias) * 10 neurons = 650 parameters
Total parameters = 100480 + 8256 + 650 = 109386 parameters, which is above 10,000. Let's reduce the size of the hidden layers to meet the requirement. We could change the number of neurons in each hidden layer to 64 and 32, respectively.

Updated model which is good on mnist dataset

In [12]:
# Define the model with reduced parameters
model = Sequential([
    Dense(64, activation='relu', input_shape=(784,)),
    Dense(32, activation='relu'),
    Dense(10, activation='softmax')
])

# Compile the model
model.compile(optimizer='adam', loss='categorical_crossentropy', metrics=['accuracy'])

# Print the model summary
model.summary()


Model: "sequential_2"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_6 (Dense)             (None, 64)                50240     
                                                                 
 dense_7 (Dense)             (None, 32)                2080      
                                                                 
 dense_8 (Dense)             (None, 10)                330       
                                                                 
Total params: 52650 (205.66 KB)
Trainable params: 52650 (205.66 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________
